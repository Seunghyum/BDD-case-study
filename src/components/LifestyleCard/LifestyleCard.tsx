import React, { ReactElement } from 'react';
import styled from 'styled-components';

import BookmarkIcon from '~components/buttons/BookmarkButton';

const CardBlock = styled.div`
  flex: 0 0 25%;
  padding: 0 10px 30px 10px;
  cursor: pointer;

  @media (max-width: 768px) {
    flex: 0 0 50%;
  }
`;

const Profile = styled.div`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  margin-bottom: 10px;
`;

const ProfileImage = styled.img`
  width: 36px;
  height: 36px;
  object-fit: contain;
  border-radius: 999px;
`;

const Nickname = styled.div`
  font-weight: bold;
  margin-left: 10px;
  color: 'rgba(0, 0, 0, 0.74)';
`;

const Content = styled.div`
  position: relative;
`;

const ContentImage = styled.img`
  width: 100%;
  border-radius: 10px;
`;

const BookmarkButton = styled.button`
  position: absolute;
  right: 10px;
  bottom: 10px;
  background: none;
  border: none;
  cursor: pointer;
`;

type CardItemProps = {
  imageUrl: string;
  nickname: string;
  profileImageUrl: string;
  isBookmarked: boolean;
  onClick: (event: React.MouseEvent<HTMLInputElement>) => void;
};

export const testId = 'bookmark';

function LifestyleCard({ imageUrl, nickname, profileImageUrl, isBookmarked, onClick }: CardItemProps): ReactElement {
  return (
    <CardBlock>
      <Profile>
        <ProfileImage src={profileImageUrl} alt="profile thumbnail" />
        <Nickname>{nickname}</Nickname>
      </Profile>
      <Content>
        <ContentImage src={imageUrl} alt="content thumbnail" />
        <BookmarkButton type="button" data-testid={testId} onClick={onClick}>
          <BookmarkIcon isBookmarked={isBookmarked} />
        </BookmarkButton>
      </Content>
    </CardBlock>
  );
}

export default LifestyleCard;
