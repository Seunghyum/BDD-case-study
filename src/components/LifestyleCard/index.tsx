export { default } from './LifestyleCard';
export { BOOKMARK_TEST_ID } from './LifestyleCard';
