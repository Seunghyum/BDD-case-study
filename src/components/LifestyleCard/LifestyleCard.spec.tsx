import React from 'react';
import { fireEvent, getByAltText, getByTestId, getByText, render } from '@testing-library/react';
import LifestyleCard, { testId } from './LifestyleCard';

describe('Feature : LifestyleCard 컴포넌트로 유저의 라이프스타일을 카드로 보여준다', () => {
  describe('Scenario : 입력값에 따라 정상적으로 렌더링 되어야한다', () => {
    it(`
      Given : imageUrl, nickname, profileImageUrl, isBookmarked, onClick을 설정할 수 있다.
      When : imageUrl, nickname, profileImageUrl을 설정할 수 있다.
      Then : 설정한 값들이 정상적으로 DOM Element에 렌더링된다
    `, () => {
      const testData = {
        imageUrl: 'test1.jpg',
        nickname: '테스트 유저',
        profileImageUrl: 'test2.jpg',
        isBookmarked: true,
        onClick: () => '',
      };

      const { container } = render(
        <LifestyleCard
          imageUrl={testData.imageUrl}
          nickname={testData.nickname}
          profileImageUrl={testData.profileImageUrl}
          isBookmarked={testData.isBookmarked}
          onClick={testData.onClick}
        />,
      );

      expect(container.firstChild).toMatchSnapshot();
      expect(getByText(container, testData.nickname)).toBeTruthy();
      expect(getByAltText(container, testData.profileImageUrl)).toBeTruthy();
      expect(getByAltText(container, testData.imageUrl)).toBeTruthy();
    });
  });

  describe('Scenario : 클릭시 설정한 콜백을 실행시켜야한다', () => {
    it(`
    Given : imageUrl, nickname, profileImageUrl, isBookmarked, onClick을 설정할 수 있다.
    When : onClick에 콜백 함수를 설정할 수 있다.
    Then : 설정한 콜백이 정상적으로 작동한다.
  `, () => {
      let isClicked = false;
      const handleClick = () => {
        isClicked = true;
      };

      const { container } = render(
        <LifestyleCard
          imageUrl="test.jpg"
          nickname="테스트 유저"
          profileImageUrl="test2.jpg"
          isBookmarked
          onClick={handleClick}
        />,
      );
      fireEvent(
        getByTestId(container, testId),
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        }),
      );
      expect(isClicked).toBeTruthy();
    });
  });
});
