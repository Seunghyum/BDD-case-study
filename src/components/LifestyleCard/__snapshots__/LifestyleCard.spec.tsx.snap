// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LifestyleCard renders LifestyleCard component 1`] = `
<div
  class="sc-gsTCUz cwkpTi"
>
  <div
    class="sc-dlfnbm kKmhwS"
  >
    <img
      alt="profile thumbnail"
      class="sc-hKgILt ebZgFx"
      src="test2.jpg"
    />
    <div
      class="sc-eCssSg kKIFdq"
    >
      테스트 유저
    </div>
  </div>
  <div
    class="sc-jSgupP dtZhcH"
  >
    <img
      alt="content thumbnail"
      class="sc-gKsewC fhZKnu"
      src="test1.jpg"
    />
    <button
      class="sc-iBPRYJ bEsZG"
      data-testid="bookmark"
      type="button"
    >
      <div
        class="sc-bdfBwQ fHaqba"
      >
        <img
          alt="bookmark-icon"
          class="bookmark-icon"
          src=""
        />
      </div>
    </button>
  </div>
</div>
`;
